### PR TITLE
[Stats Revamp] VoiceOver improvements

### LINKED ISSUE #19727: [iOS] Stats: VoiceOver issues
- When no cards are shown on the Insights tab, the “Add stats cards” button is shown but isn’t selectable via VoiceOver.
- On the Views & Visitors card, the “Weeks >” button is inaccessible by VoiceOver.
- On the Manage Stats Cards screen, when adding a card to the ACTIVE CARDS list, VoiceOver jumps seemingly randomly to focus on other cards in the list.
- After the app returns to the Insights tab after a card is added, VoiceOver reads the “Add Stats Card” button instead of reading the card that was added (or not reading anything).

### PR DESCRIPTION
**Pull request by staskus**
Fixes #19727

## Description

- [Made Add Stats Card button visible for voiceover in No Results View](https://github.com/wordpress-mobile/WordPress-iOS/commit/790a638dfff62bf857cebce443efae481390f64b) by using labels stack view as a parent element
- [When adding cards, reading appropriate hints depending on the row being in inactive or active section](https://github.com/wordpress-mobile/WordPress-iOS/commit/fac90e554ceb3649942a5d2bd7ca6243b18d83fd)
- [Reselected row for voiceover when a card is moved from active to inactive section](https://github.com/wordpress-mobile/WordPress-iOS/pull/19773/commits/dae28bfadb9add4822b04357c9d4db3514abf2d7) by using `moveRow`, `insertRow`, `deleteRow` APIs, that voiceover correctly interprets and jumps to the correct cell

Not reproducible:
- On the Views & Visitors card, the “Weeks >” button is inaccessible by VoiceOver
- After the app returns to the Insights tab after a card is added, VoiceOver reads the “Add Stats Card” button instead of reading the card that was added (or not reading anything).

## Testing instructions

See #19727

## Regression Notes

1. Potential unintended areas of impact

None

2. What I did to test those areas of impact (or what existing automated tests I relied on)

None

3. What automated tests I added (or what prevented me from doing so)

None

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### No results view action button visible for voiceover
https://user-images.githubusercontent.com/4062343/207381693-ad87095c-d6f8-4d99-9511-403750bf8a76.mov

### Week button visible for voiceover (issue that wasn't reproduced)
https://user-images.githubusercontent.com/4062343/207381743-fa0b86e2-5e6b-4a90-92aa-a28d46a3dbf3.mov

### Reading appropriate hint depending on a row belonging in an inactive or active section
https://user-images.githubusercontent.com/4062343/207381822-ac83b500-43bd-455c-ab40-b7851ef47c5e.mov

### Reselecting row for voiceover when a card is moved from active to inactive section

https://user-images.githubusercontent.com/4062343/207818131-1e9006f2-c232-4a15-999d-794c17f5491a.mov
https://user-images.githubusercontent.com/4062343/207818177-580cb690-442d-4ab6-9ab7-0722c35ad950.mp4

### Coming back to an empty Insights view - back button selected  (issue that wasn't reproduced)

https://user-images.githubusercontent.com/4062343/207382115-a0be529d-cc95-4815-be17-0e09af0cf3c4.mov

### CODE REVIEW

**Review by guarani — COMMENTED** [review]
> When no cards are shown on the Insights tab, the “Add stats cards” button is shown but isn’t selectable via VoiceOver.

This looks good now, I can select this button using VoiceOver 👍 

> On the Views & Visitors card, the “Weeks >” button is inaccessible by VoiceOver.

Like you, I couldn't reproduce this either, I tested on `trunk`.

> On the Manage Stats Cards screen, when adding a card to the ACTIVE CARDS list, VoiceOver jumps seemingly randomly to focus on other cards in the list.

This still happens for me. When moving cards from one section to another, VoiceOver focus switches to a seemingly random card before landing on the correct one. I see this in your "Reselecting row for voiceover when a card is moved from active to inactive section" screen recording. When removing Total Likes, the focus temporarily switches to "All-Time" before landing on Total Likes.

> After the app returns to the Insights tab after a card is added, VoiceOver reads the “Add Stats Card” button instead of reading the card that was added (or not reading anything).

Again, like you, I couldn't reproduce this either, I tested on `trunk`.

**Comment by staskus** [comment]
Thanks for testing, @guarani!

> This still happens for me. When moving cards from one section to another, VoiceOver focus switches to a seemingly random card before landing on the correct one.

Yes, that is the correct observation. My previous fix would only refocus on the correct row after the switch. The issue happens because we call `tableView.reloadData` on the whole table, and voiceover refocuses on one of the reused cells. It does not know that we moved that particular cell to a different section. To fix this issue properly, we need to manually move/delete/insert rows, which allows voiceover to understand what to focus on.

Check https://github.com/wordpress-mobile/WordPress-iOS/pull/19773/commits/dae28bfadb9add4822b04357c9d4db3514abf2d7 commit and new videos under `Reselecting row for voiceover when a card is moved from active to inactive section` section. 

Bonus: since we are moving the rows, we get a nice animation.

**Comment by sla8c** [comment]
@staskus I've tested this and I've found an issue - after moving a row from the active cards section to the inactive section the accessibility hint is not updated and the voiceover still says "Stat is already displayed in Insights".

This is also an issue when moving from inactive section to active section (voiceover says "Select to add this Stat to insights")

I've made a recording: 

https://user-images.githubusercontent.com/88816724/207640792-2c8ccd59-887a-4f7b-a697-aa694ff89d69.mp4

**Comment by staskus** [comment]
> @staskus I've tested this and I've found an issue - after moving a row from the active cards section to the inactive section the accessibility hint is not updated and the voiceover still says "Stat is already displayed in Insights".

Great catch! iOS doesn't reload data when using `.moveRow` and therefore the hint is not updated. I'll make a fix, it may require not using `.moveRow` (and animation that comes with it) if I don't find a solution.

**Comment by guarani** [comment]
Thanks for the update @staskus!

> I'll make a fix, it may require not using .moveRow (and animation that comes with it) if I don't find a solution.

I saw you re-requested a review from me and then @sla8c reviewed and found an issue. Would you prefer I go ahead with reviewing or hold off until you say this is ready?

**Comment by staskus** [comment]
> Would you prefer I go ahead with reviewing or hold off until you say this is ready?

@guarani I'll re-request once it's ready.

**Comment by staskus** [comment]
> @staskus I've tested this and I've found an issue - after moving a row from the active cards section to the inactive section the accessibility hint is not updated and the voiceover still says "Stat is already displayed in Insights"

I made a fix, reloading the data of the moved row so the accessibility data would update.

File: WordPress/Classes/ViewRelated/Stats/Insights/Insights Management/InsightsManagementViewController.swift
Comment (guarani) [inline WordPress/Classes/ViewRelated/Stats/Insights/Insights Management/InsightsManagementViewController.swift:327]: nitpick: Since the row was just added to the data source, it seems like `?? (insightsShown.count - 1)` will never be called. Can you explain why this is needed?

**Review by guarani — APPROVED** [review]
This is working great! I verified the following:
- [the issue that @sla8c caught earlier is fixed](https://github.com/wordpress-mobile/WordPress-iOS/pull/19773#issuecomment-1351659684)
- the rows read their state correctly when tapped
- rows that move between active/inactive sections retain their VoiceOver selection rectangle
- the rows in the inactive section retain their ordering
- placeholder rows behave as expected

I like the new table view updates which look smooth and play nicely with VoiceOver as you mentioned above.

File: WordPress/Classes/ViewRelated/Stats/Insights/Insights Management/InsightsManagementViewController.swift
Comment (staskus) [reply WordPress/Classes/ViewRelated/Stats/Insights/Insights Management/InsightsManagementViewController.swift:327]: Thanks for the review, @guarani!

Good observation, it was to avoid having an `optional` and moving the cell to the bottom just-in-case. Although this just-in-case scenario should **never** happen. 

We could use `guard let` and add logging + `tableView.reloadData` to indicate that scenario shouldn't be happening.

Comment (guarani) [reply WordPress/Classes/ViewRelated/Stats/Insights/Insights Management/InsightsManagementViewController.swift:327]: Gotcha! I don't have any strong feelings about this, happy for you to do what you think is best (even if that's to leave it as-is).

Comment (staskus) [reply WordPress/Classes/ViewRelated/Stats/Insights/Insights Management/InsightsManagementViewController.swift:327]: I think it's reasonable to leave it as it is. Additional `guard` won't improve the solution much.

**Pull request merged**
